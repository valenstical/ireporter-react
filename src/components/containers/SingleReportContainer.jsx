import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { Redirect } from 'react-router-dom';
import { getUser, getReports } from '../../selectors';
import SingleReport from '../views/SingleReport';
import updateReport from '../../actions/updateReportAction';

/**
 * Container component for the single report view
 * @export
 * @class EditReortContainer
 * @extends {Component}
 */
class SingleReportContainer extends Component {
  state = {
    report: {},
    refresh: false,
  }

  componentDidMount() {
    const { location: { pathname }, reports } = this.props;
    const pathAsArray = pathname.split('/');
    const reportId = pathAsArray[3];
    const report = reports.data.find(item => item.id.toString() === reportId.toString());
    this.populateReport(report);
  }

  handleUpdateStatus = async (event) => {
    const status = event.target.value;
    const { updateReport: processUpdateReport } = this.props;
    const { report } = this.state;

    const payload = {
      route: `${report.type}s/${report.id}/status`,
      status
    };
    await processUpdateReport(payload);
    this.setState({
      refresh: true
    });
  }

  populateReport = (report) => {
    this.setState({
      report
    });
  }

  render() {
    const { user } = this.props;
    const { report, refresh } = this.state;
    if (!user.isLoggedIn) return <Redirect to="/login" />;
    if (!report) return <Redirect to="/404" />;
    if (refresh) return <Redirect to="/login" />;
    return (
      report.status
        ? (
          <SingleReport
          report={report}
          handleUpdateStatus={this.handleUpdateStatus}
          isAdmin={user.isAdmin}
          />
        )
        : null
    );
  }
}

SingleReportContainer.propTypes = {
  user: PropTypes.object.isRequired,
  reports: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  updateReport: PropTypes.func.isRequired,

};

const mapStateToProps = createStructuredSelector(
  {
    user: getUser,
    reports: getReports
  }
);

const mapDispatchToProps = dispatch => ({
  updateReport: payload => dispatch(updateReport(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SingleReportContainer);
