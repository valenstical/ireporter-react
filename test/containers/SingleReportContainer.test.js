import React from 'react';
import setup from '../setup';
import SingleReportContainer from '../../src/components/containers/SingleReportContainer';


describe('<SingleReportContainer>', () => {
  it('should render without crashing', () => {
    const wrapper = setup(<SingleReportContainer location={{}} />);
    expect(wrapper).toMatchSnapshot();
  });
});
