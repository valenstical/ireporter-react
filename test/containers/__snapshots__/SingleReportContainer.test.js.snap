// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleReportContainer> should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "REPLACE",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/login",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(SingleReportContainer)
        location={
          Object {
            "pathname": "a/b/45",
            "reports": Object {},
          }
        }
      >
        <SingleReportContainer
          location={
            Object {
              "pathname": "a/b/45",
              "reports": Object {},
            }
          }
          reports={
            Object {
              "data": Array [],
              "isBusy": false,
              "success": false,
            }
          }
          updateReport={[Function]}
          user={
            Object {
              "isBusy": false,
              "message": Array [],
              "success": false,
              "token": "",
            }
          }
        >
          <Redirect
            push={false}
            to="/login"
          />
        </SingleReportContainer>
      </Connect(SingleReportContainer)>
    </Router>
  </BrowserRouter>
</Provider>
`;
